### ISSUE TITLE
Piping |start error traces through :talk truncates the relevant information

### ISSUE THREAD

**Issue by ohAitch** [urbit/urbit#584]
```
       ~doznec[drum]:_ford: core /~doznec/home/~2015.11.25..00.29.54..afbe/ape
       ~doznec[drum]:_ford: fade /~doznec/home/~2015.11.25..00.29.54..afbe/ape
       ~doznec[drum]: file not found
```

Maybe rectify revisions better? The general problem of having more information than can fit on a line of output, however, seems poorly solved by either truncation or `\/`.

**Comment by philipcmonk** [comment]
Just as annoying is the fact that you have to have talk running to get error messages.  It makes developing on moons confusing.

**Comment by ohAitch** [comment]
@cgyarvin, was the original rationale that you could start apps remotely and talk would work as a message bus to get errors back to your porch?

**Comment by cgyarvin** [comment]
The proper fix to get a line that fits is not to truncate the line at the
end, but collapse it in the middle.

Really, errors on moons should be sent not to the moon but to the planet.
Basically, moons should not be on talk directly.

**Comment by ohAitch** [comment]
By "middle", do you mean "ford"? "~doznec/home/{compile-date}"? That line
doesn't really contain completely irrelevant information, though perhaps
the date could stand to be truncated slightly; certainly not irrelevant
information that can be algorithmically determined for any tape.

**Comment by cgyarvin** [comment]
I simply mean that if you have a long line of text and you don't know
what's in it, cutting bytes out of the middle is likely to work better than
truncating.

**Comment by ohAitch** [comment]
Better, sure, but does not resolve the core issue of "expanding stack
traces line by line is unreasonable, and the data in them does not fit on
one line of screen, esp when prefixed with talk[system] boilerplate".

**Comment by cgyarvin** [comment]
Yes.  The correct solution is that the stack trace is shown by activating
the message, not sent line by line.

**Comment by ohAitch** [comment]
Right, %tang is a %fat type.

**Comment by philipcmonk** [comment]
> Really, errors on moons should be sent not to the moon but to the planet.
> Basically, moons should not be on talk directly.

Um, no?  I wish to use moons for dev, and I want to see y'all's messages on my planet.  I don't want to sully up my planet with error messages, and I want to be able to use a moon for dev work offline.

**Comment by cgyarvin** [comment]
Suggest a better solution, then -- what do you think is the right
architecture?

**Comment by philipcmonk** [comment]
Remind me again why we don't simply crash?

**Comment by ohAitch** [comment]
+1

**Comment by ohAitch** [comment]
Using moons as a proxy for planets should be an /option/, but my conception
of the primary purpose of moons is IoT and fakenet type
subordinate independent agents.

**Comment by cgyarvin** [comment]
Simply crash, meaning what?  The process exits?  Are we solving the same
problem?

Constraints: I want to be developing and talking on a moon -- this should
be the normal setup.  My moon is local, my planet is in the cloud, I'm on
line.   I need to see all kinds of notifications and errors in my talk
stream.  Should they all make the process exit?  I *certainly* need to see
private messages sent to my planet.  I don't want two talk connections open
for this.

Are any of these constraints erroneous?  If not, how do you deal with them?

Given that the normal case is online operation, it seems reasonable that "I
want ~wictuc-folrex to see this message" translates into "send this message
to ~wictuc-folrex".  Even if the cause of the message is "an error happened
on ~wictuc-folrex's moon."  If you establish that this is the normal case,
and the offline case is a corner case, you can work out how to design for
the corner case.  But I'm not sure we're there yet.

**Comment by philipcmonk** [comment]
By "simply crash" I mean the old behavior.

> I need to see all kinds of notifications and errors in my talk stream.

This isn't obvious to me.

How about another constraint:  I need my errors to *not* be spammed across all my moons and my planet or else development slows.

The most important constraint, though, is that when I fire up a moon and try to run an app that doesn't compile, I see something somewhere without having to remember to start up talk in some way.

**Comment by ohAitch** [comment]
And relatedly, when I try to send a message to an app that isn't started, I
see something somewhere; though that is a separate issue perhaps

**Comment by cgyarvin** [comment]
In this case, I don't have a problem dropping a %crud back on arvo and
feeding the error back to the console ("just crash" / "old behavior").  In
fact, I'm not sure when or why this behavior was changed.  If we have a
direct causal path to the console, we should use it.

But there are other cases where we don't have a direct causal path to the
console -- eg, update causes a sync conflict.  So we still have the
notification problem.

In general, your talk stream is certainly your notification stream.  Your
moon does a console login to your planet's :talk, unless someone's changed
it since I configured it last, so you don't have to remember to start up
talk in any way.  You do have to remember not to stop it.

I'm not quite sure what you mean by "or else development slows" -- can you
clarify?

I feel like some of the motivation for not wanting to run :talk is the
distraction effect of urbit-meta.  This is very true.  Nothing wrong with a
;leave urbit-meta in this case.  If you do this, you're still getting
private messages et al, which you probably still want.

**Comment by philipcmonk** [comment]
Talk on moons does not get autostarted.

**Comment by juped** [comment]
It was removed a few months ago - I think I might have complained about it for some reason, though I don't remember the details. It can probably go back.

**Comment by cgyarvin** [comment]
I see you're right, although I could swear I both implemented this and
documented it.  Not sure I want to dig into the logs and research what
happened.

**Comment by cgyarvin** [comment]
In any case, these changes are trivial -- the action item is to figure out
what we *should* be doing.

**Comment by juped** [comment]
https://github.com/urbit/urbit/blob/master/urb/zod/lib/drum.hoon#L75

On a very cursory inspection, that *is* what's in drum. Why isn't it working?

**Comment by juped** [comment]
Oh, it's *linked* to, in `++deft-fish`, but never actually *started* in `++deft-apes`.

**Comment by juped** [comment]
(The culprit is one "C. Guy Yarvin".)

**Comment by ohAitch** [comment]
That makes sense, since the one it's linked to is the parent.

**Comment by juped** [comment]
Yeah, it does make sense. But why isn't it working?

**Comment by ohAitch** [comment]
Also I could have sworn I added %ask on that list for ~doznec

**Comment by ault011** [comment]
I don't really understand this that well, but basically this caused me a huge problem when the stack trace informed me that my syntax error was in the incorrect file.

```
: /~zod/home/42/ape/gmail/hoon
          ~zod[drum]:_ford: core /~zod/home/~2016.2.8..23.14.04..4b77/ape/gmai
          ~zod[drum]:_ford: fade /~zod/home/~2016.2.8..23.14.04..4b77/ape/gmai
-----------[75]
          ~zod[drum]: syntax error: 23 2
: /~zod/home/43/ape/gmail/hoon
          ~zod[drum]:_ford: core /~zod/home/~2016.2.8..23.14.44..c664/ape/gmai
          ~zod[drum]:_ford: wilt /~zod/home/~2016.2.8..23.14.44..c664/ape/gmai
          ~zod[drum]:_ford: fade /~zod/home/~2016.2.8..23.14.44..c664/ape/gmai
          ~zod[drum]: syntax error: 21 22
-----------[80]
          ~zod: (scan "    ' next)
> (scan "    " next)
{1 2}
'syntax-error'
ford: build failed
--------------| ;77
? 77
0v1.l7t6o.tft1b.s9arh.43aq0.asnk5 at ~2016.2.8..23.14.44..c664 ~zod to :
[drum]: ford: wilt /~zod/home/~2016.2.8..23.14.44..c664/ape/gmail/split
```

I was getting the the `21 22` error message as shown above, but that line on the gmail app did not have a column `22` (it was, however, a column that I had just been working on, so I believed an error on my part was probable). Then Anton came over and still couldn't figure it out.

It wasn't until I told him 'hey, I also just worked on this other file, maybe our error messages are messed up really badly and this is the error', that he remembered that there actually was an issue here. Only then did he do the `;77` to print out the *actual* file where the error was occurring.

Anyways, this needs to be fixed badly because, as we have now seen, it has the potential to throw smart people--even Anton--off.

**Comment by ohAitch** [comment]
Closed by urbit/arvo#79; it is survived by #712.
